src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+        """Get number of samples per pulse.
+
+        Digitizer traces are acquired by train, by may contain the data
+        of several pulses, with each pulse covering a certain number of
+        samples depending on the digitizer and machine repetition rates.
+
+        This method may be called with any combination of arguments to
+        determine the number of samples per pulse, with order of
+        precendence in definition order. Calling with no arguments tries
+        to use pulse information the component was initialized with.
+
+        Args:
+            repetition_rate (float): Pulse repetition rate in Hz.
+            pulse_period (int or float): Pulse period in PPT units (int)
+                or seconds (float).
+            pulse_ids (ArrayLike, optional): Pulse IDs.
+            fractional (bool, optional): Whether to round to possible
+                EuXFEL repetition rates (default) or return the full
+                fractional value.
+
+        Returns:
+            (int or float) Number of samples per pulse, float if
+                fractional=True.
+        """
+
+        if isinstance(pulse_period, float):
+            # Float are interpreted as pulse period in seconds, convert
+            # to units of 4.5 MHz.
+            pulse_period = AdqChannel._bunch_repetition_rate * pulse_period
+
+        elif not isinstance(pulse_period, int):
+            # Any non-int value assumes the pulse period has to be
+            # determined from other values.
+
+            if repetition_rate is not None:
+                pulse_period = AdqChannel._bunch_repetition_rate \
+                    / repetition_rate
+            elif pulse_ids is None and self._pulses is not None:
+                pulse_ids = self._pulses.peek_pulse_ids()
+
+            if pulse_ids is not None:
+                # May either be passed directly or come from pulses
+                # component.
+                try:
+                    pulse_period = set(pulse_ids[1:] - pulse_ids[:-1])
+                except IndexError:
+                    raise ValueError('two or more pulses requires to infer '
+                                     'pulse period') from None
+
+                if len(pulse_period) > 1:
+                    raise ValueError('more than one period between pulse IDs')\
+                        from None
+
+                pulse_period = int(pulse_period.pop())
+
+        if pulse_period is None:
+            raise ValueError('must pass either pulse_period, repetition_rate, '
+                             'pulse_ids or initialize component with pulse '
+                             'information') from None
+
+        if not fractional:
+            pulse_period = int(round(pulse_period, 1))
+
+        return self._clock_ratio * pulse_period
+
+    def preprocess(self, data, out=None, dtype=None,
+                   cm_period=8, baseline=np.s_[:1000], baselevel=None):
+        """Preprocess raw digitizer data.
```

Comment:
Thanks for the feedback!

I'm certainly not really happy about some of the interface choices, so I hope we can find something better with more eyes. That being said, I have the feeling I may have forgot some context or not stress it enough, so I'll first have to disagree on a few points:

* I understand where you're coming from with detector data, but for digitizers *raw* data indeed has a very particular meaning. ADQ digitizers (and the same names are used for FastADC) have four distinct output modes for the signal they acquire:
  * `raw`: Continuously sampled signal for up to few ms.
  * `zero`: Fixed length slices of the signal as response to a threshold, can be outside of `raw` trigger
  * `apd`: Integrated signal per pulse
  * `peak`: Position of _peaks_ (what constitutes a peak is up to the FPGA here, and geared towards certain use cases)
 
  These may be enabled in any combination and are quite explicit both in device configuration and data structure. The component should expose all of these output modes on equal footing, and while I see the unfortunate collision with *raw*, I feel strongly to use the same terms. That's why I moved away from the first draft still having `.ndarray()` etc, it was exactly a `raw`-focused view I want to remedy. 

* Yes, I am fully convinced that any look on raw data should include common mode correction. As a matter of fact, after contacting the vendor (Teledyne) about my findings I got confirmation it is a known consequence of the parallel readout architecture. I actually hope that this component helps in spreading this correction to any user. That being said, it will not be relevant in every use case depending on the SNR - but when averaging a trace, the common mode *noise* can grow beyond the single trace random noise (!) depending on the board. It made the difference between us seeing ${\rm Xe}^{43+}$ or not.

* I understand the not careful checking of `preprocess_kw` in that typos or wrong values can easily be passed to the initializer before they get checked during actual use. I see the issue in that, my motivation to move them to the initializer was early user feedback I got that they do not want to constantly repeat such settings when loading data. From there I was afraid of having too many arguments to functions, but that's likely preferred to collating them into unchecked dicts. It being a potential no-op is intentional. As mentioned essentially always you should correct for common mode, and in 99% of cases you want to pull some baseline to zero to make signals interpretable.

* I cannot follow how going by *chunk* in the EXtra-data sense does not its current objective of efficiently loading *corrected* data into memory. The common mode correction require a type change to work (even for a baseline pull you *should* go `float`), and the goal is exactly to avoid a full copy of the entire data, which was the case in the very early versions. Going by chunk will at most load one file into memory, which currently stands at 14G for a full 16 channel ADQ412 covering 50s time. A typical run either taken with REMI or XAS will be a couple of minutes, so the goal of this method is to load 50G-200G into memory and it profiled favorably for these cases against loading - casting - correcting. I intentionally did not fully parallelize this method yet, and likely splitting them arbitrarily would be of benefit here.

My own as well as the SQS requirement (covering various TOF and diode timing applications) for this component is to have single-line methods to efficiently obtain *usable* ***raw*** data as well as discriminated edges, not force users to always chain the same 2-3 methods in 99% of cases. This necessitates offering a common mode correction or at least baseline pull and different output layouts as requested. Apart from the method of pulse separation I daresay XAS has the same needs, and even for pulse separation I suspect the same given my EXtra demo some time ago produced identical results.

Based on your feedback and some inspiration from the masked data PR:

* To work around the naming issue, introduce distinct classes from each output mode, i.e. `AdqRawChannel` and `AdqZeroChannel` in the future. These can carry much more generic methods like `.ndarray()`, `.series()`, 
*`.edges()`, though it still leaves by train vs by pulse to be resolved.

* They could be convenient created with an `AdqDigitizer` by something like `AdqDigitizer(run).raw(cm_period=4).ndarray()`, i.e. trying to distribute the configurability via subtypes.

* Expand `preprocess_kw` as suggested in the initializer of output modes that require it and validate it. It will not be needed for other output modes like `AdqPeakChannel`. 

* Likely we can get rid of the word `preprocess` entirely in the arguments, and as suggested split up the methods for manual processing. If you insist we can introduce boolean like `apply_cm`, though I suspect the current design should be understandable given documentation (like I already attempted) and should also only affect very few users.

Reply:
Thanks for the context. :+1: 

> ...the goal is exactly to avoid a full copy of the entire data, which was the case in the very early versions. Going by chunk will at most load one file into memory...

If you have virtual overview files, though, one chunk as seen by EXtra-data can be a contiguous set of data in the virtual overview file, i.e. the entire run. So if you have the RunValidator set up, as we do for FXE, you're back to a full copy of all the data.

> If you insist we can introduce boolean like apply_cm

I don't insist, but maybe in the docstring of the 'raw' methods we could add something like 'to disable this step, create the component with cm_period=0'. Otherwise people have to jump around the docs to figure that out.

How much time do we have for playing with the design? I know you wanted to use this soon.

Reply:
> If you have virtual overview files, though, one chunk as seen by EXtra-data can be a contiguous set of data in the virtual overview file

I did not consider that, thanks. My first reaction when reading your comment was something like "a chunk but at most N rows", that should work in all situations then.

> How much time do we have for playing with the design? I know you wanted to use this soon.

I want to deliver something to next week's users, with a particular focus on edge discrimination by pulse. These users would also like to asap see code examples to prepare. Hence I would put emphasis on API design much more than implementation (e.g. the case here about chunking). If you feel we cannot nail down a solid future-forward design without more time, my substitute would be some standalone code installed into our environment for that purpose that we stop supporting at some point (or replace by calls to EXtra).

Reply:
To be honest, I think it's already pretty solid, and will be even more so with some tweaks along the lines we've discussed. :+1: 

I think we could come up with something more elegant if we took more time, but I think delivering useful things to users should take precedence over refining APIs, and we're not going to get it perfect however long we take.